ext/opcache/ZendAccelerator.c

```diff
@@ -2451,6 +2451,298 @@ static zend_class_entry* zend_accel_inheritance_cache_add(zend_class_entry *ce,
+			break;
+		case ZEND_FFI_TYPE_ARRAY:
+			new_type->array.type = accel_ffi_persist_type_copy(ptr, new_type->array.type);
+			break;
+		case ZEND_FFI_TYPE_STRUCT:
+			ht = &new_type->record.fields;
+			GC_SET_REFCOUNT(ht, 2);
+			GC_ADD_FLAGS(ht, IS_ARRAY_IMMUTABLE);
+			HT_FLAGS(ht) |= HASH_FLAG_STATIC_KEYS;
+			ht->pDestructor = NULL;
+			ht->nInternalPointer = 0;
+			if (HT_IS_PACKED(ht)) {
+				memcpy(*ptr, ht->arPacked, HT_PACKED_USED_SIZE(ht));
+				ht->arPacked = (zval*)(*ptr);
+				(*ptr) = (void*)((char*)(*ptr) + HT_PACKED_USED_SIZE(ht));
+			} else {
+				memcpy(*ptr, HT_GET_DATA_ADDR(ht), HT_USED_SIZE(ht));
+				HT_SET_DATA_ADDR(ht, (Bucket*)(*ptr));
+				(*ptr) = (void*)((char*)(*ptr) + HT_USED_SIZE(ht));
+			}
+			ZEND_HASH_FOREACH_BUCKET(&new_type->record.fields, b) {
+			    f = (zend_ffi_field*)(*ptr);
+				memcpy(f, Z_PTR(b->val), sizeof(zend_ffi_field));
+				Z_PTR(b->val) = f;
+				(*ptr) = (void*)((char*)(*ptr) + sizeof(zend_ffi_field));
+				f->type = accel_ffi_persist_type_copy(ptr, f->type);
+			} ZEND_HASH_FOREACH_END();
+			break;
+		case ZEND_FFI_TYPE_FUNC:
+			if (new_type->func.ret_type) {
+				new_type->func.ret_type = accel_ffi_persist_type_copy(ptr, new_type->func.ret_type);
+			}
+			if (new_type->func.args) {
+				ht = (*ptr);
+				memcpy(ht, new_type->func.args, sizeof(HashTable));
+				(*ptr) = (void*)((char*)(*ptr) + sizeof(HashTable));
+				new_type->func.args = ht;
+				GC_SET_REFCOUNT(ht, 2);
+				GC_ADD_FLAGS(ht, IS_ARRAY_IMMUTABLE);
+				HT_FLAGS(ht) |= HASH_FLAG_STATIC_KEYS;
+				ht->pDestructor = NULL;
+				ht->nInternalPointer = 0;
+				ZEND_ASSERT(HT_IS_PACKED(ht));
+				memcpy(*ptr, ht->arPacked, HT_PACKED_USED_SIZE(ht));
+				ht->arPacked = (zval*)(*ptr);
+				(*ptr) = (void*)((char*)(*ptr) + HT_PACKED_USED_SIZE(ht));
+				ZEND_HASH_PACKED_FOREACH_VAL(ht, zv) {
+					Z_PTR_P(zv) = accel_ffi_persist_type_copy(ptr, (zend_ffi_type*)Z_PTR_P(zv));
+				} ZEND_HASH_FOREACH_END();
+			}
+			break;
+		default:
+			break;
+	}
+
+	return new_type;
+}
+
+static zend_ffi_dcl* accel_ffi_persist_type(zend_ffi_dcl *dcl)
+{
+	void *ptr;
+	zend_ffi_dcl *new_dcl;
+	ssize_t size;
+
+	zend_shared_alloc_init_xlat_table();
+	size = accel_ffi_persist_type_calc(ZEND_FFI_TYPE(dcl->type));
+	zend_shared_alloc_destroy_xlat_table();
+
+	if (size == -1) {
+		return NULL;
+	}
+	ptr = zend_shared_alloc(sizeof(zend_ffi_dcl) + size);
+	if (!ptr) {
+		return NULL;
+	}
+
+	zend_shared_alloc_init_xlat_table();
+
+	new_dcl = (zend_ffi_dcl*)ptr;
+	ptr = (void*)((char*)ptr + sizeof(zend_ffi_dcl));
+	memcpy(new_dcl, dcl, sizeof(zend_ffi_dcl));
+	new_dcl->type = accel_ffi_persist_type_copy(&ptr, new_dcl->type);
+
+	zend_shared_alloc_destroy_xlat_table();
+
+	return new_dcl;
+}
+
+static zend_result accel_ffi_cache_type_get(zend_string *str, zend_ffi_dcl *dcl)
+{
+	str = accel_find_interned_string(str);
+	if (str && (str->gc.u.type_info & IS_STR_FFI_TYPE)) {
```

Comment:
You may need to include the cdef in the lookup key in some way, as `str` may depend on it.

E.g.:

``` php
$cdef = FFI::cdef("typedef char test;");
$cdata = $cdef->new("test");

$cdef = FFI::cdef("typedef int test;");
$cdata = $cdef->new("test");
```

An other possible issue is that this could lead to a high number of cached types due to types like `$cdef->new('char[' . $len . ']')`.

Reply:
> You may need to include the cdef in the lookup key in some way, as `str` may depend on it.

yes. One of the PHPT tests already catches this problem. The work is in progress...

Reply:
Now it's possible to cache types with the same name declared in different ``contexts``. Conflicts are resolved through a linked list.